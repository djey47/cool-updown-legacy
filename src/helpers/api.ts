/**
 * Provides utility functions to handle Cool-updown API requests and responses
 */

import { ApiRequest } from "../model/models";

/**
 * @param request API request from Express middleware
 * @returns server identifier which is passed through a path parameter
 */
export function extractServerIdentifier(request: ApiRequest) {
  return request.params.serverId;
}
