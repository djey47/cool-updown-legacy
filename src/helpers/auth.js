/* @flow */

const basicAuth = require('express-basic-auth');

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
function initBasicAuth(app, isEnabled, userName, password) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

module.exports = {
  initBasicAuth,
};
