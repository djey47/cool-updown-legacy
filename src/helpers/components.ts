/*
 * Misc. page components
 */

/**
 * @returns contents in a pragraph then a link in a next one
 */
export function withBackLink(contents: string, linkUrl: string, linkText: string) {
    return `
    <p class="backlink-contents">${contents}</p>
    <p class="backlink-link"><a href="${linkUrl}">${linkText}</a></p>
    `;
}
