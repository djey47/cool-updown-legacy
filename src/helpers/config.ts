import config from 'config';
import { TypedResponse } from "../model/express";
import { ServerConfig } from '../model/models';
import messages from '../resources/messages';

/**
 * 
 * @param res 
 * @param serverId 
 * @returns 
 */
export function retrieveServerConfiguration(res: TypedResponse<string>, serverId: string) {
  const serversConfiguration = config.get('servers');
  const serverConfiguration = serversConfiguration[serverId] as ServerConfig;

  if (!serverConfiguration) {
    res.status(400).send(messages.errors.invalidArg);
  }

  return serverConfiguration;
}
