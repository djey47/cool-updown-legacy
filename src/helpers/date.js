const MS_IN_1_S = 1000;
const MS_IN_1_M = MS_IN_1_S * 60;
const MS_IN_1_H = MS_IN_1_M * 60;
const MS_IN_1_D = MS_IN_1_H * 24;

/**
 * @returns Object with fields days, hours, minutes, seconds, milliseconds
 */
function getTimeDetails(durationMs) {
  let milliseconds = durationMs;

  const days = Math.floor(milliseconds / MS_IN_1_D);
  milliseconds %= MS_IN_1_D;

  const hours = Math.floor(milliseconds / MS_IN_1_H);
  milliseconds %= MS_IN_1_H;

  const minutes = Math.floor(milliseconds / MS_IN_1_M);
  milliseconds %= MS_IN_1_M;

  const seconds = Math.floor(milliseconds / MS_IN_1_S);
  milliseconds %= MS_IN_1_S;

  return {
    milliseconds,
    seconds,
    minutes,
    hours,
    days,
  };
}

module.exports = {
  getTimeDetails,
};
