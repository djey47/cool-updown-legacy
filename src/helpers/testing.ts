/**
 * Unit testing support functions
 */

/**
 * @param expressResponseMock 
 * @returns an Express default response (with mocks) for unit tests
 */
export const generateDefaultResponse = (expressResponseMock) => ({
  status: c => expressResponseMock.statusMock(c),
  send: msg => expressResponseMock.sendMock(msg),
  json: ()  => expressResponseMock.jsonMock(),
});    
