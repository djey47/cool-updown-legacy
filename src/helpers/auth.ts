import basicAuth from 'express-basic-auth';
import config from 'config';
import { readFile } from 'fs/promises';

interface App {
  use: (plugin: unknown) => unknown;
}

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
export function initBasicAuth(app: App, isEnabled: boolean, userName: string, password: string) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

/**
 * @private
 * @return Promise which resolves to String,
 * with the contents of SSH private key as per provided keyPath in configuration
 */
export async function readPrivateKey() {
  const privateKeyPath = config.get('server.keyPath') as string;
  const privateKey = await readFile(privateKeyPath, 'utf-8');

  // console.log('services::getPrivateKey', { privateKeyPath, privateKey });

  return privateKey as string;
}
