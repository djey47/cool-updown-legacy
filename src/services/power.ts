import { NodeSSH } from 'node-ssh';
import wol, { WakeOptions } from 'wake_on_lan';
import logger from '../helpers/logger';
import messages from '../resources/messages';
import { readPrivateKey } from '../helpers/auth';
import { ApiRequest, AppState } from '../model/models';
import { TypedResponse } from '../model/express';
import { validateInputParameters } from '../helpers/api';
import { retrieveServerConfiguration } from '../helpers/config';
import { generatePage } from '../helpers/page';

const ssh = new NodeSSH();

/**
 * Handles ON request for all servers
 */
export function on(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  // TODO
  res.send(generatePage('Not implemented yet'));
}

/**
 * Handles OFF request for all servers
 */
export function off(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  // TODO
  res.send(generatePage('Not implemented yet'));
}

/**
 * Handles ON request
 */
export function onServer(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  const { serverId } = validateInputParameters(req as ApiRequest, res);
  const serverConfiguration = retrieveServerConfiguration(res, serverId);

  const macAddress = serverConfiguration.network?.macAddress || undefined;
  const broadcastAddress = serverConfiguration.network?.broadcastIpAddress || undefined;
  const options: WakeOptions = {
    address: broadcastAddress,
  };
  wol.wake(macAddress, options, (error) => {
    if (error) {
      logger.error(`(on-server:${serverId}) ${messages.wakeKO} - ${JSON.stringify(error, null, '  ')}`);

      if (res) res.status(500).send(messages.status.kayo);
    } else {
      logger.info(`(on-server:${serverId}) ${messages.wakeOK}`);

      // console.log('onServer', { appState }, appState.servers);

      const serverState = appState.servers[serverId];
      if (!serverState.startedAt) {
        serverState.startedAt = new Date(Date.now());
        serverState.stoppedAt = undefined;
      }

      if (res) res.send(generatePage(messages.status.okay));
    }
  });
}

/**
 * Handles OFF request
 */
export async function offServer(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  const { serverId } = validateInputParameters(req as ApiRequest, res);
  const serverConfiguration = retrieveServerConfiguration(res, serverId);
  
  const host = serverConfiguration.network?.hostname || undefined;
  const port = serverConfiguration.ssh?.port;
  const username = serverConfiguration.ssh?.user;
  const password = serverConfiguration.ssh?.password;
  const command = serverConfiguration.ssh?.offCommand;
  const privateKeyPath = serverConfiguration.ssh?.keyPath || undefined;

  try {
    const privateKey = await readPrivateKey(privateKeyPath);
    await ssh.connect({
      host,
      port,
      username,
      privateKey,
    });

    logger.info(`(off-server:${serverId}) ${messages.sshOK}`);

    const { stdout, stderr, code } = await ssh.execCommand(command, { stdin: `${password}\n` });
    logger.log('debug', `(off-server:${serverId}) STDOUT: ${stdout}`);
    logger.log('debug', `(off-server:${serverId}) STDERR: ${stderr}`);

    if (code !== 0) throw stderr;

    const serverState = appState.servers[serverId];
    if (!serverState.stoppedAt) {
      serverState.startedAt = undefined;
      serverState.stoppedAt = new Date(Date.now());
    }

    if (res) res.send(generatePage(messages.status.okay));
  } catch (err) {
    logger.error(`(off-server:${serverId}) ${messages.sshKO} - ${JSON.stringify(err, null, '  ')}`);

    if (res) res.status(500).send(generatePage(messages.status.kayo));
  } finally {
    ssh.dispose();
  }
}
