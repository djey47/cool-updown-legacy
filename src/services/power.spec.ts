import globalMocks from '../../config/jest/globalMocks';
import resetMocks from '../../config/jest/resetMocks';
import { generateDefaultResponse } from '../helpers/testing';
import { off, on } from './power';

const { expressResponseMock, nodeFSMock, nodesshMock, wakeonlanMock } = globalMocks;

const appState = {
  isScheduleEnabled: true,
  startedAt: new Date('June 12, 2018 13:14:00Z'),
};

const res = generateDefaultResponse(expressResponseMock);

describe('power services', () => {
  describe('on service', () => {
    beforeEach(() => {
      resetMocks();

      // WOL OK
      wakeonlanMock.wake.mockImplementation((a, o, f) => {
        f();
      });

      expressResponseMock.statusMock.mockImplementation(() => expressResponseMock);
    });

    it('should invoke wol and generate correct response on success', () => {
      // given
      const appStateWithServerStopTime = {
        ...appState,
        serverStartedAt: null,
        serverStoppedAt: new Date(),
      };

      // when
      on(undefined, res, appStateWithServerStopTime);

      // then
      expect(wakeonlanMock.wake).toHaveBeenCalled();
      expect(wakeonlanMock.wake.mock.calls[0][0]).toEqual('FF:FF:FF:FF:FF:FF:FF:FF');
      expect(wakeonlanMock.wake.mock.calls[0][1]).toEqual({ address: '255.255.255.255' });
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
      expect(appStateWithServerStopTime.serverStartedAt.getTime()).toEqual(1528812840000);
      expect(appStateWithServerStopTime.serverStoppedAt).toBeUndefined();
    });

    it('should not alter server start time when already defined', () => {
      // given
      const appStateWithServerStartTime = {
        ...appState,
        serverStartedAt: new Date(),
      };

      // when
      on(undefined, res, appStateWithServerStartTime);

      // then
      expect(appStateWithServerStartTime.serverStartedAt).toBeDefined();
      expect(appStateWithServerStartTime.serverStartedAt.getTime()).not.toEqual(1528812840000);
    });

    it('should invoke wol and generate correct response on failure', () => {
      // given
      // WOL KO
      wakeonlanMock.wake.mockImplementation((a, o, f) => {
        f({});
      });
      const appStateWithServerStopTime = {
        ...appState,
        serverStartedAt: undefined,
        serverStoppedAt: new Date(),
      };

      // when
      on(undefined, res, appStateWithServerStopTime);

      // then
      expect(wakeonlanMock.wake).toHaveBeenCalled();
      expect(expressResponseMock.statusMock).toHaveBeenCalledWith(500);
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
      expect(appStateWithServerStopTime.serverStartedAt).toBeUndefined();
      expect(appStateWithServerStopTime.serverStoppedAt).not.toBeUndefined();
    });
  });

  describe('off service', () => {
    beforeEach(() => {
      resetMocks();

      // Async file reader OK with private key contents
      nodeFSMock.readFile.mockResolvedValue('=== PRIVATE KEY ===');

      expressResponseMock.statusMock.mockImplementation(() => expressResponseMock);

      nodesshMock.execCommand.mockResolvedValue({ stdout: '', stderr: '', code: 0 });
    });

    it('should invoke SSH and generate correct response on success', async () => {
      // given
      const appStateWithServerStartTime = {
        ...appState,
        serverStartedAt: new Date(),
        serverStoppedAt: undefined,
      };

      // when
      await off(undefined, res, appStateWithServerStartTime);

      // then
      expect(nodesshMock.connect).toHaveBeenCalledWith({
        host: 'Host Name',
        port: 22,
        privateKey: '=== PRIVATE KEY ===',
        username: 'User',
      });
      expect(nodesshMock.execCommand).toHaveBeenCalledWith('OFF Command');
      expect(nodesshMock.dispose).toHaveBeenCalled();
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
      expect(appStateWithServerStartTime.serverStartedAt).toBeUndefined();
      expect(appStateWithServerStartTime.serverStoppedAt.getTime()).toEqual(1528812840000);
    });

    it('should not alter server stop time when already defined', async () => {
      // given
      const appStateWithServerStopTime = {
        ...appState,
        serverStoppedAt: new Date(),
      };

      // when
      await off(undefined, res, appStateWithServerStopTime);

      // then
      expect(appStateWithServerStopTime.serverStoppedAt).toBeDefined();
      expect(appStateWithServerStopTime.serverStoppedAt.getTime()).not.toEqual(1528812840000);
    });

    it('should invoke SSH and generate correct response on failure (promise rejection)', async () => {
      // given
      // ExecCommand KO (rejection)
      nodesshMock.execCommand.mockImplementation(() => Promise.reject());
      const appStateWithServerStartTime = {
        ...appState,
        serverStartedAt: new Date(),
        serverStoppedAt: null,
      };

      // when
      await off(undefined, res, appStateWithServerStartTime);

      // then
      expect(nodesshMock.connect).toHaveBeenCalled();
      expect(nodesshMock.execCommand).toHaveBeenCalled();
      expect(nodesshMock.dispose).toHaveBeenCalled();
      expect(expressResponseMock.statusMock).toHaveBeenCalledWith(500);
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
      expect(appStateWithServerStartTime.serverStartedAt).not.toBeUndefined();
      expect(appStateWithServerStartTime.serverStoppedAt).toBeNull();
    });

    it('should invoke SSH and generate correct response on command failure', async () => {
      // given
      // ExecCommand KO (non 0 exit code)
      nodesshMock.execCommand.mockImplementation(() => Promise.resolve({ stdin: '', stdout: '', code: 1 }));

      // when
      await off(undefined, res, appState);

      // then
      expect(nodesshMock.connect).toHaveBeenCalled();
      expect(nodesshMock.execCommand).toHaveBeenCalled();
      expect(nodesshMock.dispose).toHaveBeenCalled();
      expect(expressResponseMock.statusMock).toHaveBeenCalledWith(500);
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should invoke SSH and generate correct response on connect failure', async () => {
      // given
      // Connect KO
      nodesshMock.connect.mockImplementation(() => Promise.reject());

      // when
      await off(undefined, res, appState);

      // then
      expect(nodesshMock.connect).toHaveBeenCalled();
      expect(nodesshMock.execCommand).not.toHaveBeenCalled();
      expect(nodesshMock.dispose).toHaveBeenCalled();
      expect(expressResponseMock.statusMock).toHaveBeenCalledWith(500);
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
