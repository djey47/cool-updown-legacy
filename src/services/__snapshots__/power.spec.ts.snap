// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`power services off service for single server should invoke SSH and generate correct response on command failure 1`] = `"<page-shared />KO"`;

exports[`power services off service for single server should invoke SSH and generate correct response on connect failure 1`] = `"<page-shared />KO"`;

exports[`power services off service for single server should invoke SSH and generate correct response on failure (promise rejection) 1`] = `"<page-shared />KO"`;

exports[`power services off service for single server should invoke SSH and generate correct response on success 1`] = `"<page-shared />OK"`;

exports[`power services on service for single server should invoke wol and generate correct response on failure 1`] = `"KO"`;

exports[`power services on service for single server should invoke wol and generate correct response on success 1`] = `"<page-shared />OK"`;
